File: backend/app/api/v1/social.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import get_current_user, get_current_user_optional
from app.core.responses import success
from app.models.user import User
from app.repositories import follow_repository
from app.services import follow_service

router = APIRouter()


@router.get("/users/{user_id}/follow-status")
async def get_follow_status(
    user_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User | None = Depends(get_current_user_optional),
):
    is_following = await follow_repository.is_following(
        db,
        current_user.id if current_user else None,
        user_id,
    )
    return success({"is_following": is_following, "user_id": str(user_id)})


@router.post("/users/{user_id}/follow")
async def post_follow_user(
    user_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await follow_service.follow_user(db, user_id, current_user.id)
    return success(result)


@router.delete("/users/{user_id}/follow")
async def delete_follow_user(
    user_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await follow_service.unfollow_user(db, user_id, current_user.id)
    return success(result)


@router.get("/users/{user_id}/connections")
async def get_user_connections(
    user_id: UUID,
    type: str = Query(default="followers"),
    limit: int = Query(default=50, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User | None = Depends(get_current_user_optional),
):
    connection_type = "following" if type == "following" else "followers"
    result = await follow_service.get_connections(
        db,
        user_id,
        connection_type,
        limit,
        current_user.id if current_user else None,
    )
    return success(result)

File: frontend/src/api/social.ts
import { apiClient } from '@/lib/apiClient';

export async function followUser(userId: string): Promise<{ following: boolean }> {
  const res = await apiClient.post(`/users/${userId}/follow`);
  return res.data;
}

export async function unfollowUser(userId: string): Promise<{ following: boolean }> {
  const res = await apiClient.delete(`/users/${userId}/follow`);
  return res.data;
}

export async function getFollowStatus(
  userId: string,
): Promise<{ is_following: boolean; user_id: string }> {
  const res = await apiClient.get(`/users/${userId}/follow-status`);
  return res.data;
}

export async function getConnections(
  userId: string,
  type: 'followers' | 'following',
  limit = 50,
): Promise<{
  type: string;
  total: number;
  users: Array<{
    id: string;
    username: string;
    display_name: string | null;
    photo_url: string | null;
    is_self: boolean;
    is_following: boolean;
  }>;
}> {
  const res = await apiClient.get(
    `/users/${userId}/connections?type=${type}&limit=${limit}`,
  );
  return res.data;
}
